### ISSUE TITLE
[HELP, BUG] RWStructuredBuffer usage Dynamic fails on all but Vulkan backend.

### ISSUE THREAD

**Issue by thargy** [mellinoe/veldrid#80]
Hi @mellinoe,

I've have finally managed to get a compute shader test work end-to-end for [`ShaderGen`](https://github.com/mellinoe/ShaderGen) using `Veldrid`.  To make it work I need to read back the data from a `RWStructuredBuffer` after running the compute shader.  Unfortunately, I couldn't find an example of how to read back from a `RWStructuredBuffer`, and although I came up with a solution that compiles and runs it only succeeds on the Vulkan (`Glsl450`) backend and fails with different errors on ever other backend.

The code to run the Compute Shader can be found [here](https://github.com/thargy/ShaderGen/blob/f283e437771e1205395f14399f666b7975969636/src/ShaderGen.Tests/ShaderBuiltinsTests.cs#L112-L187), and the Shader itself, and buffer can be found [here](https://github.com/thargy/ShaderGen/blob/f283e437771e1205395f14399f666b7975969636/src/ShaderGen.Tests/TestAssets/ShaderBuiltinsComputeTest.cs#L18), and makes use of [this struct](https://github.com/thargy/ShaderGen/blob/f283e437771e1205395f14399f666b7975969636/src/ShaderGen.Tests/TestAssets/ComputeShaderParameters.cs).  The compute shader currently [only tests](https://github.com/thargy/ShaderGen/blob/f283e437771e1205395f14399f666b7975969636/src/ShaderGen.Tests/TestAssets/ShaderBuiltinsComputeTest.cs#L37-L52) the four overloads for `Abs()`, but now it's working I can easily extend to test all the implemented overloads on the backends.  The tests run the shader code on the CPU (in parallel) and then on the GPU and compares the results automatically, outputting any differences, which makes it incredibly easy to identify and correct inconsistencies between CPU/GPU and different backends.  As this is testing from C# -> Transpilation -> Compilation -> Execution, it is truly an end to end test; and now the principle is established it should be a small step to enhance to support Vertex/Fragment shader testing by outputing to a bitmap (although this is unlikely to be executable on the CPU and so will need to compare GPU outputs directly).

I'm currently having to create a hidden window, which doesn't seem to cause any issues, but I'm wondering if the code can be reduced for the Compute Shader tests at least so it doesn't rely on SDL2?  That may make it more acceptable in CI environments?

The new `FactAttribute` overrides also auto-skip `GraphicsBackends` that are unavailable, so this _shouldn't_ kill the CI server, as it's testing for the presence of the compiler and support for the backend.

Anyway, outside of Vulkan here are the outputs of the failing tests:

### Glsl330

> Veldrid.VeldridException : Error(s) were encountered during the execution of OpenGL commands. See InnerException for more information.
> ---- System.IndexOutOfRangeException : Index was outside the bounds of the array.
>    at Veldrid.OpenGL.OpenGLGraphicsDevice.ExecutionThread.CheckExceptions()
>    at Veldrid.OpenGL.OpenGLGraphicsDevice.ExecutionThread.WaitForIdle()
>    at Veldrid.GraphicsDevice.WaitForIdle()
>    at ShaderGen.Tests.ShaderBuiltinsTests.TestShaderBuiltins(ToolChain toolChain) in D:\Source Control\ShaderGen\src\ShaderGen.Tests\ShaderBuiltinsTests.cs:line 178
>    at ShaderGen.Tests.ShaderBuiltinsTests.TestShaderBuiltins_Glsl330() in D:\Source Control\ShaderGen\src\ShaderGen.Tests\ShaderBuiltinsTests.cs:line 36
> ----- Inner Stack Trace -----
>    at Veldrid.OpenGL.OpenGLShader.CreateGLResources()
>    at Veldrid.OpenGL.OpenGLShader.EnsureResourcesCreated()
>    at Veldrid.OpenGL.OpenGLPipeline.CreateComputeGLResources()
>    at Veldrid.OpenGL.OpenGLPipeline.CreateGLResources()
>    at Veldrid.OpenGL.OpenGLCommandExecutor.ActivateComputePipeline()
>    at Veldrid.OpenGL.NoAllocEntryList.OpenGLNoAllocCommandEntryList.ExecuteAll(OpenGLCommandExecutor executor)
>    at Veldrid.OpenGL.OpenGLGraphicsDevice.ExecutionThread.ExecuteWorkItem(ExecutionThreadWorkItem workItem)
> 
> Generated shader set for Glsl330 backend.
> Compiled Compute Shader from set "TestShaders.ShaderBuiltinsComputeTest.CS"!
> Generated random parameters for 4 methods.
> Created compute pipeline.
> 

### GlslEs300

> Veldrid.VeldridException : Error(s) were encountered during the execution of OpenGL commands. See InnerException for more information.
> ---- System.IndexOutOfRangeException : Index was outside the bounds of the array.
>    at Veldrid.OpenGL.OpenGLGraphicsDevice.ExecutionThread.CheckExceptions()
>    at Veldrid.OpenGL.OpenGLGraphicsDevice.ExecutionThread.WaitForIdle()
>    at Veldrid.GraphicsDevice.WaitForIdle()
>    at ShaderGen.Tests.ShaderBuiltinsTests.TestShaderBuiltins(ToolChain toolChain) in D:\Source Control\ShaderGen\src\ShaderGen.Tests\ShaderBuiltinsTests.cs:line 178
>    at ShaderGen.Tests.ShaderBuiltinsTests.TestShaderBuiltins_GlslEs300() in D:\Source Control\ShaderGen\src\ShaderGen.Tests\ShaderBuiltinsTests.cs:line 32
> ----- Inner Stack Trace -----
>    at Veldrid.OpenGL.OpenGLShader.CreateGLResources()
>    at Veldrid.OpenGL.OpenGLShader.EnsureResourcesCreated()
>    at Veldrid.OpenGL.OpenGLPipeline.CreateComputeGLResources()
>    at Veldrid.OpenGL.OpenGLPipeline.CreateGLResources()
>    at Veldrid.OpenGL.OpenGLCommandExecutor.ActivateComputePipeline()
>    at Veldrid.OpenGL.NoAllocEntryList.OpenGLNoAllocCommandEntryList.ExecuteAll(OpenGLCommandExecutor executor)
>    at Veldrid.OpenGL.OpenGLGraphicsDevice.ExecutionThread.ExecuteWorkItem(ExecutionThreadWorkItem workItem)
> 
> Generated shader set for GlslEs300 backend.
> Compiled Compute Shader from set "TestShaders.ShaderBuiltinsComputeTest.CS"!
> Generated random parameters for 4 methods.
> Created compute pipeline.
> 

### Hlsl

> SharpDX.SharpDXException : HRESULT: [0x80070057], Module: [General], ApiCode: [E_INVALIDARG/Invalid Arguments], Message: The parameter is incorrect.
> 
>    at SharpDX.Result.CheckError()
>    at SharpDX.Direct3D11.Device.CreateBuffer(BufferDescription& descRef, Nullable`1 initialDataRef, Buffer bufferOut)
>    at Veldrid.D3D11.D3D11Buffer..ctor(Device device, UInt32 sizeInBytes, BufferUsage usage, UInt32 structureByteStride)
>    at Veldrid.D3D11.D3D11ResourceFactory.CreateBufferCore(BufferDescription& description)
>    at Veldrid.ResourceFactory.CreateBuffer(BufferDescription description)
>    at ShaderGen.Tests.ShaderBuiltinsTests.TestShaderBuiltins(ToolChain toolChain) in D:\Source Control\ShaderGen\src\ShaderGen.Tests\ShaderBuiltinsTests.cs:line 134
>    at ShaderGen.Tests.ShaderBuiltinsTests.TestShaderBuiltins_Hlsl() in D:\Source Control\ShaderGen\src\ShaderGen.Tests\ShaderBuiltinsTests.cs:line 44
> 
> Generated shader set for Hlsl backend.
> Compiled Compute Shader from set "TestShaders.ShaderBuiltinsComputeTest.CS"!
> Generated random parameters for 4 methods.
> 

Unfortunately, I can't test Metal as I don't have a Mac.

You'll note that Direct3D fails on trying to write to the buffer using
```csharp
                    // Update parameter buffer
                    graphicsDevice.UpdateBuffer(inOutBuffer, 0, gpuParameters);
```

Whereas the two OpenGL pipelines (not Vulkan) fail later, when executing `graphicsDevice.WaitForIdle();`  The above highlights how easy it now is to find and show inconsistencies between backends.

I've tried to figure out the solution, and my best guess is that only Vulkan doesn't mind me setting the Buffer Usage flag for both the buffer to `BufferUsage.StructuredBufferReadWrite | BufferUsage.Dynamic` (as can be seen [here](https://github.com/thargy/ShaderGen/blob/b032cf013d2e074aedda27cbc72d8f2b12bf3516/src/ShaderGen.Tests/ShaderBuiltinsTests.cs#L137)).  However, the only way I could find to read back from the buffer was to use `GraphicsDevice.Map` and that requires either the `BufferUsage.Dynamic` or `BufferUsage.Staging` flag to be set, and the latter cannot be set with `BufferUsage.StructuredBufferReadWrite` which is also required (note, not a lot of helpful error messages to figure that lot out).  Note, however, all of the failures occur before mapping the resource to read it back.

I can at least now continue work on mellinoe/ShaderGen#65 in relation to Vulkan support, however, I lost a lot of time fixing bugs as I went along and I'm going to be very busy on other work from tomorrow.  For now I've implemented the methods in `ShaderBuiltins` (see [here](https://github.com/thargy/ShaderGen/blob/b032cf013d2e074aedda27cbc72d8f2b12bf3516/src/ShaderGen.Primitives/ShaderBuiltins.cs#L91-L316)), and in Glsl330 (see [here](https://github.com/thargy/ShaderGen/blob/b032cf013d2e074aedda27cbc72d8f2b12bf3516/src/ShaderGen/Glsl/Glsl330KnownFunctions.cs#L19-L73)).

The branch has quite a lot of bug fixes and enhancements, and it passes all existing tests (with the exception of the above), so it wouldn't do much harm to pull it into main if you're still working on a beta and wish to play with the new features.  I will try to get it all completed soon, and generate a pull request but work commitments will slow me down going forward.

**Comment by mellinoe** [comment]
* `BufferUsage.StructuredBufferReadWrite` + `Dynamic` should actually be disallowed earlier in the creation process, since it isn't possible on D3D11. The D3D11 code is actually failing when it tries to create the buffer rather than update it. I thought that I had actually made that change in the past (restricting the BufferUsage flags), but it seems that it will still let you try that combination.

* Reading back data from buffers is done through `BufferUsage.Staging` buffers. Lots of the tests here in this repo read back data from buffers, so you can take a look at those to see how it works.

* The OpenGL backend does a lot of work in a separate thread, so the timing of errors can be different, depending on what order calls are made. `WaitForIdle` does a full flush of all buffered commands in the execution thread, so oftentimes errors will be propagated out at that time. As for why there's an error in that place, I'm not 100% sure, but looking over the code it's potentially because the shader bytes passed in are null.

**Comment by mellinoe** [comment]
> I'm currently having to create a hidden window, which doesn't seem to cause any issues, but I'm wondering if the code can be reduced for the Compute Shader tests at least so it doesn't rely on SDL2? That may make it more acceptable in CI environments?

All of the backends except for OpenGL support creating a headless GraphicsDevice, without any window or main swapchain. These are the `GraphicsDevice.Create()` overloads that only take a `GraphicsDeviceOptions`.

For example:
* [CreateD3D11(GraphicsDeviceOptions)](https://mellinoe.github.io/veldrid-docs/api/Veldrid.GraphicsDevice.html#Veldrid_GraphicsDevice_CreateD3D11_Veldrid_GraphicsDeviceOptions_)
* [CreateVulkan(GraphicsDeviceOptions)](https://mellinoe.github.io/veldrid-docs/api/Veldrid.GraphicsDevice.html#Veldrid_GraphicsDevice_CreateVulkan_Veldrid_GraphicsDeviceOptions_)

OpenGL doesn't have any headless overloads at the moment, so a hidden window is needed there.

**Comment by thargy** [comment]
THANKS!  That's what I needed.  I had seen the staging before for Textures but was confused.  Adding a staging buffer resolved the issues.

The OpenGL failures were due to empty arrays (bug in the way I implemented that fallback to raw code when no compilation output is created), I've fixed that too.

The tests are all currently working and passing for all backends.  I'm away for a few days, but when I get back I'll try to finish the Maths functions for all backends as it should be easy now.
